driver/asana_driver.py

```diff
@@ -0,0 +1,185 @@
+    
+    return dt
+
+
+class AsanaDriver(object):
+
+    def __init__(self, butter_user_id, datasource_user_id):
+        self.butter_user_id = butter_user_id
+        self.datasource_user_id = datasource_user_id
+        
+        # pull api credentials
+        personal_access_token = os.environ['ASANA_ACCESS_TOKEN'] # TODO: handle authentication better
+        
+        # create api client
+        self.client = asana.Client.access_token(personal_access_token)
+        
+        # pull initial data
+        try:
+            self.me = self.client.users.me()
+            self.workspaces = self.me['workspaces']
+            self.projects = []
+            for w in self.workspaces:
+                for p in self.client.projects.find_all({ 'workspace': w['id'] }):
+                    #p['workspace'] = w
+                    self.projects.append(p)
+                    pass
+                pass
+        except asana.error.NoAuthorizationError, e:
+            raise AuthRevokedError( e.message )
+        except asana.error.RateLimitEnforcedError, e:
+            raise RateLimitError( e.message )
+       
+        # configure client
+        asana.Client.DEFAULTS['page_size'] = 1000
+        return
+
+    def iter_tasks(self):
+        """
+        Returns a list of all tasks.
+        """
+        
+        for p in self.projects:
+            for t in self.client.tasks.find_all({ 'project': p['id'] }):
+                t['project'] = p
+                yield t
+                pass
+            pass
+        
+        return
+
+    def retrieve_data(self, doc):
+        """
+        Used to download content for a single doc. Typically this involves
+        passing the doc `external_id` attribute to the datasource SDK to
+        initialize a download. It is not mandatory that the only function
+        this performs is downloading. Gmail, for example, discovers
+        all the emails in a thread using the root email within `retrieve_data`.
+
+        :param doc: A dictionary describing a document.
+
+        :return `RetrieveDataResult` with the appropriate values populated.
+        """
+        external_id = doc['external_id']
+        subtype = doc.get('subtype')
+        
+        try:
+            if subtype == 'attachment':
+                attachment = self.client.attachments.find_by_id(attachment=external_id)
+                
+                url = attachment['download_url']
+                
+                r = requests.get(url)
+                
+                return RetrieveDataResult(data=r.content) # TODO: check if this okay for potentially large files
+            elif subtype is not None:
+                raise Exception( 'unexpected subtype: ' + repr(s) )
+            else:
+                task = self.client.tasks.find_by_id(task=external_id)
+                
+                return RetrieveDataResult(data=task['notes']) # Note: technically, a this could be empty
+        except asana.error.NotFoundError, e:
```

Comment:
Does this mean the doc is no longer within their system? (task/attachment is deleted?)